GetOldTweets3/manager/ConcurrentTweetManager.py

```diff
@@ -0,0 +1,78 @@
+from threading import Thread
+from queue import Queue
+from queue import Empty
+import copy
+
+from GetOldTweets3.manager.TweetManager import TweetManager
+
+DATE_FORMAT = "%Y-%m-%d"
+FIRST_TWEET_DATE = datetime.strptime("2006-03-21", DATE_FORMAT)
+
+
+class ConcurrentTweetManager:
+
+    @staticmethod
+    def getTweets(tweetCriteria, receiveBuffer=None, bufferLength=100, proxy=None, debug=False, worker_count=1, ordered=True):
+        # Init the queues
+        time_spans = Queue()
+        tweets = []
+        workers = []
+
+        # Split the date in smaller parts
+        since_date = datetime.strptime(tweetCriteria.since, DATE_FORMAT) if tweetCriteria.since else FIRST_TWEET_DATE
+        until_date = datetime.strptime(tweetCriteria.until, DATE_FORMAT) if tweetCriteria.until else datetime.now()
+        date_diff_per_worker = (until_date - since_date) / worker_count
+
+        # Create a TweetCriteria that can be cloned as Model by the Workers
+        criteria = copy.deepcopy(tweetCriteria)
+        criteria.setMaxTweets(criteria.maxTweets / worker_count)
+
+        # Create a time span for each of the splitted parts and a corresponding worker
+        for i in range(1, worker_count+1):
+            from_time = since_date + (i-1) * date_diff_per_worker
+            to_time = since_date + i * date_diff_per_worker
+            time_spans.put((from_time, to_time))
+
+            w = WorkerThread(copy.deepcopy(criteria), time_spans, tweets, receiveBuffer, bufferLength, proxy, debug)
+            w.start()
+            workers.append(w)
+
+        # Wait for the workers to finish and then stop them
+        time_spans.join()
+        for worker in workers:
+            worker.stop()
+
+        return sorted(tweets, key=lambda r: r.date)
+
+
+class WorkerThread(Thread):
+
+    def __init__(self, tweetCriteria, time_spans, tweets, receiveBuffer=None, bufferLength=100, proxy=None, debug=False):
+        super().__init__()
+        self.stopped = False
+        self.manager = TweetManager()
+        self.tweetCriteria = tweetCriteria
+        self.receiveBuffer = receiveBuffer
+        self.bufferLength = bufferLength
+        self.proxy = proxy
+        self.debug = debug
+        self.time_spans = time_spans
+        self.tweets = tweets
+
+    def stop(self):
+        self.stopped = True
+
+    def run(self) -> None:
+        while not self.stopped:
+            try:
+                time_span = self.time_spans.get_nowait()
+                self.tweetCriteria.setSince(datetime.strftime(time_span[0], "%Y-%m-%d"))
+                self.tweetCriteria.setUntil(datetime.strftime(time_span[1], "%Y-%m-%d"))
+                search_results = self.manager.getTweets(self.tweetCriteria, self.receiveBuffer, self.bufferLength,
+                                                        self.proxy, self.debug)
+                self.tweets.extend(search_results)
+                self.time_spans.task_done()
+            except Empty:
+                pass
```

Comment:
Maybe just `return` from function?
Can `self.stopped` be removed from the thread killing logic?

Reply:
The way it is built right now it is necesarry, so that the main thread can use join() to wait for workers to finish. But you‘re right, that can be made cleaner by just passing the time_span to the worker within the constructor and getting rid of the queue altogehter... 
Only thing we‘ll need to figure out is how the mainprocess will wait till each process is done... We could spin in a while loop like (while(not_all_threads_stopped) sleep(10ms)) or sth... Or we could use a semaphore that each worker will reduce by 1 if hes done at the end and make the main thread wait for the semaphore to be 0 to proceed.... What do you think=

Reply:
`queue.Queue.join()` is waiting for all elements in queue to be done with `self.time_spans.task_done()`.
I think the strait forward way is to:  
1. make first loop to create a proper `self.time_spans` list. 
2. make a second loop that create `worker_count` workers each of which takes the next element from `self.time_spans` in a loop and finish when nothing to retrieve from the list.

Reply:
I discovered the following which also works and wont need a queue at all:

```
        for worker in workers:
            worker.join()
```